File: 백준/Gold/1461. 도서관/도서관.py
N, M = map(int, input().split())
arr = list(map(int, input().split()))
arr.sort()
lst1 = []  # 음수 담기
lst2 = []  # 양수 담기
max_idx = 0
count = 0
for i in range(N):
    if arr[i] > 0:
        lst2.append(arr[i])
    else:
        lst1.append(arr[i])

if lst1 != [] and lst2 != []:
    if abs(lst1[0]) > lst2[-1]:  # 제일 먼 거리가 음수인 경우
        max_idx = 1
    else:  # 제일 먼 거리가 양수인 경우
        max_idx = 2

for i in range(0, len(lst1), M):
    if max_idx == 1 or max_idx == 0:  # 제일 먼 거리가 음수인 경우 or lst1만 함수가 있는 경우
        max_idx = -1
        count += abs(lst1[i])
    else:
        count += abs(lst1[i])*2

for i in range(len(lst2) - 1, -1, -M):
    if max_idx == 2 or max_idx == 0:  # 제일 먼 거리가 양수인 경우 or lst2만 함수가 있는 경우
        max_idx = -1
        count += lst2[i]
    else:
        count += lst2[i]*2
print(count)

File: 백준/Gold/1461. 도서관/도서관.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class Main {
	public static void main(String[] args) throws Exception{
		// 입력 받기
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt(); // 책 개수
		int M = sc.nextInt(); // 한번 이동할 때 들 수 있는 최대 책 개수
		int arr[] = new int[N];
		for (int i = 0; i < N; i++) {
			arr[i] = sc.nextInt();
		}
		
		Arrays.sort(arr); // 내림차순 정렬
		// 음수만 담는 배열과 양수만 담는 배열 분리
		ArrayList<Integer> minus_arr = new ArrayList<Integer>();
		ArrayList<Integer> plus_arr = new ArrayList<Integer>();
		
		// 각 배열에 값 넣기
		for (int i = 0; i < N; i++) {
			if (arr[i]> 0) {
				plus_arr.add(arr[i]);
			}
			else {
				minus_arr.add(arr[i]);
			}
		}
		int minus_size = minus_arr.size();
		int plus_size = plus_arr.size();
		int idx = 0;  // 이 친구가 핵심 키임! 1이면 음수쪽이 큼, 2면 양수쪽이 큼, 한쪽에 값이 존재하지 않으면 0
		if (minus_arr.isEmpty() == false && plus_arr.isEmpty()== false) {
			if ((-1)*minus_arr.get(0) > plus_arr.get(plus_size-1)) {
				idx = 1; // 음수쪽의 젋댓값이 더 큰 경우
			}
			else {
				idx = 2; // 양수쪽의 절댓값이 더 큰 경우
			}
		}
		int count = 0; // 결괏값
		// 음수쪽 계산하기
		for (int i = 0; i < minus_size;i+=M) {
			if (idx == 1 || idx == 0) {
				count -= minus_arr.get(i);
				idx = -1;
			}
			else {
				count -= minus_arr.get(i)*2; 
			}
		}
		// 양수쪽 계산하기
		for (int i = plus_size-1; i > -1;i-=M) {
			if (idx == 2 || idx == 0) {
				count += plus_arr.get(i);
				idx = -1;
			}
			else {
				count += plus_arr.get(i)*2; 
			}
		}
		
		System.out.println(count);
	}
}